File: src/za/co/adhd_developers/java/questions/Question26.java
package za.co.adhd_developers.java.questions;

import za.co.adhd_developers.java.tools.Question;
import za.co.adhd_developers.java.tools.TimeCheck;
import za.co.adhd_developers.java.tools.Utils;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;

/**
 * Created by Grant on 2017/08/01.
 */
public class Question26 implements Question {
    int answer = 0;
    String cycle = "";
    int limit = 1000;
    String prefix = "";

    @Override
    public void doWork() {
        BigDecimal result;
        BigDecimal one = new BigDecimal("1");
        for (int i = 999; i > 1; i--) {

            int cycleLength = this.cycle.length();
            result = one.divide(new BigDecimal(String.valueOf(i)), (i - 1) * 2, RoundingMode.DOWN);
            result = result.stripTrailingZeros();

            String number = result.toString();
            String recurring = number.substring(number.indexOf(".") + 1);
            int recurring_length = recurring.length();

            int HALF_LENGTH = Math.floorDiv(recurring_length, 2);
            if (HALF_LENGTH == 0) {
                continue;
            }

            if (HALF_LENGTH < cycleLength) {
                continue;
            }

            for (int prefixLength = 0; prefixLength <= 5; prefixLength++) {
                String prefix = "";

                String workingSet = recurring;
                if (prefixLength > 0) {
                    workingSet = recurring.substring(prefixLength);
                    prefix = recurring.substring(0,prefixLength);
                }

                int workingHALF_LENGTH = Math.floorDiv(workingSet.length(),2);

                if (workingHALF_LENGTH == 0) {
                    break;
                }

                if (workingHALF_LENGTH < cycleLength) {
                    break;
                }

                for (int copyLength = 1; copyLength <= workingHALF_LENGTH; copyLength++) {
                    String[] items = workingSet.split("(?<=\\G.{" + copyLength + "})");
                    if (items.length == 1) {
                        continue;
                    }

                    int count = 0;
                    for (String item : items) {
                        if (item.equals(items[0])) {
                            count++;
                        } else {
                            break;
                        }
                    }

                    if (count == items.length) {
                        if (cycleLength < items[0].length()) {
                            this.cycle = items[0];
                            this.answer = i;
                            copyLength = workingHALF_LENGTH + 1;
                            cycleLength = this.cycle.length();
                            this.prefix = prefix;
                        } else {
                            break;
                        }
                    }
                }
            }
        }
    }

    @Override
    public void printAnswer() {
        System.out.println("=================================");
        System.out.println("Question 26");
        System.out.println("The 1/d with the longest recurring cycle, is: " + this.answer);
        System.out.println("The cycle for 1/" + this.answer + ", is 0." + this.prefix + "(" + this.cycle + ")");
        System.out.println("=================================");
    }
}

File: src/za/co/adhd_developers/kotlin/questions/Question26.kt
package za.co.adhd_developers.kotlin.questions

import za.co.adhd_developers.kotlin.Question
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.regex.Pattern

class Question26 : Question {

    var answer = 0
    var oldLength = 0
    override fun doWork() {
        for (number in 999 downTo 2) {
            val scale = (number - 1) * 2
            if (scale / 2 < oldLength) {
                continue
            }
            val original = BigDecimal.ONE.divide(BigDecimal("$number"), scale, RoundingMode.DOWN).stripTrailingZeros().toString()

            val recurring = original.substring(2)

            var copyFromIndex = 0
            var copyLength = 1

            while (copyFromIndex <= number.minus(1)) {
                while (copyFromIndex + copyLength <= number.minus(1)) {
                    var prefex = ""
                    if (copyFromIndex > 0) {
                        prefex = recurring.substring(0, copyFromIndex)
                    }

                    val checkFor = recurring.substring(copyFromIndex, copyFromIndex + copyLength)
                    val regex = "^0.${if (prefex.isNotEmpty()) "($prefex)" else ""}($checkFor){2,}\$"

                    val matcher = Pattern.compile(regex).matcher(original)

                    if (matcher.find()) {
                        copyFromIndex = number.minus(1)

                        val newLength = checkFor.length
                        if (oldLength < newLength) {
                            oldLength = newLength
                            answer = number
                        }
                    } else {
                        copyLength++
                    }
                }
                copyFromIndex++
                copyLength = 1
            }
        }
    }

    override fun printAnswer() {
        println("=================================")
        println("Question 26")
        println("The 1/d with the longest recurring cycle, is: ${this.answer}")
        println("The length of the cycle for 1/${this.answer}, is ${this.oldLength}")
        println("=================================")
    }
}